src/Gui/View3DInventorViewer.cpp

```diff
@@ -642,35 +642,46 @@ View3DInventorViewer::~View3DInventorViewer()
     delete glAction;
 }
 
-void View3DInventorViewer::createStandardCursors(double dpr)
+// QTBUG-95434
```

Comment:
In terms of clean code this shouldn't be a macro but a proper function. Then then whole cursor handling code should be refactored and moved to its own class.

For future work:
There is also a class ToolHandler in the core system that partially handles the creation of cursor pixmaps. It would be good to refactor that as well to have all cursor related code in one class.

Reply:
I originally wrote that as a function and then moved back to the macro as it minimizes impact on other platforms. The logic is while we already need to preprocessor to deal with testing Qt versions, lets use it for all the stuff. Please note I'm able to test only on Linux and I didn't want to break anything else. I do not understand why `254ed951f924687e3dffb325288bbc0333980b9c` limited setting cursor pixel ratio to Windows only. Knowing all those details would help me to refactor code (yes, we are back to understandable commit messages).

I do not want to touch this that late in development cycle, so refactoring is out of scope now. Leaving for two weeks, so I will not be able to react to complains if I break anything I cannot test during refactor. That said, anyone is free to use this as a base for his own nice solution. If nothing happens until I return and people are okay with touching it more radical way and there is still time to give it more testing, I'm fine with that.

Reply:
> I do not understand why 254ed951f924687e3dffb325288bbc0333980b9c limited setting cursor pixel ratio to Windows only.

This is because the Win32 API works differently compared to the APIs of other platforms.

Reply:
This macro is called only once - there is no need to even declare it as function, simply inline it where it is used:

```
QPixmap View3DInventorViewer::fixupCursor(const QBitmap &bitmap, const QBitmap &mask, double dpr) const
{
#if defined(Q_OS_LINUX) && QT_VERSION < QT_VERSION_CHECK(6,6,0) && QT_VERSION >= QT_VERSION_CHECK(5,13,0)
    // QTBUG-95434
    QPixmap pixmap;
    if (qGuiApp->platformName() == QLatin1String("wayland")) {
        QImage img = bitmap.toImage();
        img.convertTo(QImage::Format_ARGB32);
        pixmap = QPixmap::fromImage(img);
    } else {
        pixmap = bitmap;
    }
#else
    QPixmap pixmap(bitmap);
#endif

    pixmap.setMask(mask);
#if defined(Q_OS_WIN32)
    pixmap.setDevicePixelRatio(dpr);
#else
    Q_UNUSED(dpr)
#endif
    return pixmap;
}
```

Reply:
> > I do not understand why [254ed95](https://github.com/FreeCAD/FreeCAD/commit/254ed951f924687e3dffb325288bbc0333980b9c) limited setting cursor pixel ratio to Windows only.
> 
> This is because the Win32 API works differently compared to the APIs of other platforms.

Yes, that is obvious from that `#if defined(Q_OS_WIN32)` in the code and there are gazilions of cursor related bugs in QT bug tracker, so some reference could be handy. I found similar solution referenced in [QTBUG-52211](https://bugreports.qt.io/browse/QTBUG-52211?focusedId=352267&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-352267) Nevermind, I won't touch this.

@kadet1090, indeed macro is used only once and it needs to `#include <QGuiApplication>` so this is just feasible way, how to do it using single ifdef. I'm aware that other workarounds adds includes unconditionally and I do not agree with that. It is implemented as a macro on purpose, to keep workaround code at single place, so it is self explaining and will be useful when removing it again.